File: reservation-service/src/main/java/com/coworking/reservations/algorithm/CacheLRU.java
package com.coworking.reservations.algorithm;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.ArrayList;

// Cache LRU desde cero: lista doblemente enlazada + hash map. O(1) get/put.
public class CacheLRU<K, V> {

    // Nodo de la lista enlazada
    private static class Nodo<K, V> {
        K clave;
        V valor;
        Nodo<K, V> prev;
        Nodo<K, V> next;

        Nodo(K clave, V valor) {
            this.clave = clave;
            this.valor = valor;
        }
    }

    private final int capacidad;
    private final Map<K, Nodo<K, V>> mapa;
    private final Nodo<K, V> head;
    private final Nodo<K, V> tail;
    private long hits = 0;
    private long misses = 0;

    public CacheLRU(int capacidad) {
        this.capacidad = capacidad;
        this.mapa = new HashMap<>();
        // Dummy head y tail
        this.head = new Nodo<>(null, null);
        this.tail = new Nodo<>(null, null);
        head.next = tail;
        tail.prev = head;
    }

    // O(1) - retorna valor o null
    public synchronized V get(K clave) {
        Nodo<K, V> nodo = mapa.get(clave);
        if (nodo == null) {
            misses++;
            return null;
        }
        moverAlFrente(nodo);
        hits++;
        return nodo.valor;
    }

    // O(1) - inserta o actualiza
    public synchronized void put(K clave, V valor) {
        Nodo<K, V> existente = mapa.get(clave);
        if (existente != null) {
            existente.valor = valor;
            moverAlFrente(existente);
            return;
        }
        Nodo<K, V> nodo = new Nodo<>(clave, valor);
        mapa.put(clave, nodo);
        insertarAlFrente(nodo);

        // Evict el menos usado
        if (mapa.size() > capacidad) {
            Nodo<K, V> lru = tail.prev;
            remover(lru);
            mapa.remove(lru.clave);
        }
    }

    public synchronized boolean eliminar(K clave) {
        Nodo<K, V> nodo = mapa.get(clave);
        if (nodo == null) return false;
        remover(nodo);
        mapa.remove(clave);
        return true;
    }

    public synchronized int tamanio() {
        return mapa.size();
    }

    public synchronized Map<String, Object> estadisticas() {
        long total = hits + misses;
        double hitRate = total > 0 ? Math.round((hits * 10000.0 / total)) / 100.0 : 0.0;
        List<String> claves = new ArrayList<>();
        for (K k : mapa.keySet()) claves.add(String.valueOf(k));
        Map<String, Object> stats = new HashMap<>();
        stats.put("capacidad", capacidad);
        stats.put("tamanio", mapa.size());
        stats.put("hits", hits);
        stats.put("misses", misses);
        stats.put("hit_rate", hitRate);
        stats.put("claves", claves);
        return stats;
    }

    // Insertar nodo justo despues de head
    private void insertarAlFrente(Nodo<K, V> nodo) {
        nodo.next = head.next;
        nodo.prev = head;
        head.next.prev = nodo;
        head.next = nodo;
    }

    private void remover(Nodo<K, V> nodo) {
        nodo.prev.next = nodo.next;
        nodo.next.prev = nodo.prev;
    }

    private void moverAlFrente(Nodo<K, V> nodo) {
        remover(nodo);
        insertarAlFrente(nodo);
    }
}

File: billing_service/src/estructuras/cache_lru.js
// Cache LRU con lista doblemente enlazada + hash map. O(1) get/put.

class Nodo {
  constructor(clave, valor) {
    this.clave = clave;
    this.valor = valor;
    this.prev  = null;
    this.next  = null;
  }
}

class CacheLRU {
  constructor(capacidad = 100) {
    this.capacidad = capacidad;
    this.mapa = new Map();
    // Dummy head y tail para simplificar inserciones
    this.head = new Nodo(null, null);
    this.tail = new Nodo(null, null);
    this.head.next = this.tail;
    this.tail.prev = this.head;
    this.hits = 0;
    this.misses = 0;
  }

  // Insertar nodo justo despues de head
  _insertar_al_frente(nodo) {
    nodo.next = this.head.next;
    nodo.prev = this.head;
    this.head.next.prev = nodo;
    this.head.next = nodo;
  }

  // Remover nodo de la lista
  _remover(nodo) {
    nodo.prev.next = nodo.next;
    nodo.next.prev = nodo.prev;
  }

  // Mover nodo existente al frente
  _mover_al_frente(nodo) {
    this._remover(nodo);
    this._insertar_al_frente(nodo);
  }

  // O(1) - retorna valor o null
  get(clave) {
    const nodo = this.mapa.get(clave);
    if (!nodo) {
      this.misses++;
      return null;
    }
    this._mover_al_frente(nodo);
    this.hits++;
    return nodo.valor;
  }

  // O(1) - inserta o actualiza
  put(clave, valor) {
    const existente = this.mapa.get(clave);
    if (existente) {
      existente.valor = valor;
      this._mover_al_frente(existente);
      return;
    }
    const nodo = new Nodo(clave, valor);
    this.mapa.set(clave, nodo);
    this._insertar_al_frente(nodo);

    // Evict el menos usado si excede capacidad
    if (this.mapa.size > this.capacidad) {
      const lru = this.tail.prev;
      this._remover(lru);
      this.mapa.delete(lru.clave);
    }
  }

  eliminar(clave) {
    const nodo = this.mapa.get(clave);
    if (!nodo) return false;
    this._remover(nodo);
    this.mapa.delete(clave);
    return true;
  }

  tamanio() {
    return this.mapa.size;
  }

  estadisticas() {
    const total = this.hits + this.misses;
    return {
      capacidad:  this.capacidad,
      tamanio:    this.mapa.size,
      hits:       this.hits,
      misses:     this.misses,
      hit_rate:   total > 0 ? Math.round((this.hits / total) * 10000) / 100 : 0,
      claves:     Array.from(this.mapa.keys()),
    };
  }

  // Vaciar completamente el cache (resetea contadores tambien)
  vaciar() {
    this.mapa.clear();
    this.head.next = this.tail;
    this.tail.prev = this.head;
    this.hits = 0;
    this.misses = 0;
  }
}

// Singletons compartidos
const cache_usuarios = new CacheLRU(100);
const cache_reportes = new CacheLRU(50);

module.exports = { CacheLRU, cache_usuarios, cache_reportes };
